src/lib.rs

```diff
@@ -0,0 +1,389 @@
+                        }
+                    }
+                    ptn == text
+                }
+                None => ptn == text,
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn defaults() {
+        fn helper(ptn: &str, text: &str) -> bool {
+            FMatcher::new(ptn).unwrap().matches(text).is_ok()
+        }
+        assert!(helper("", ""));
+        assert!(helper("a", "a"));
+        assert!(!helper("a", "ab"));
+        assert!(helper("...", ""));
+        assert!(helper("...", "a"));
+        assert!(helper("...", "a\nb"));
+        assert!(helper("...\na", "a"));
+        assert!(helper("...\na\n...", "a"));
+        assert!(helper("a\n...", "a"));
+        assert!(helper("a\n...\nd", "a\nd"));
+        assert!(helper("a\n...\nd", "a\nb\nc\nd"));
+        assert!(!helper("a\n...\nd", "a\nb\nc"));
+        assert!(helper("a\n...\nc\n...\ne", "a\nb\nc\nd\ne"));
+        assert!(helper("a\n...\n...b", "a\nb"));
+        assert!(helper("a\n...\nb...", "a\nb"));
+        assert!(helper("a\n...\nb...", "a\nbc"));
+        assert!(helper("a\nb...", "a\nbc"));
+        assert!(!helper("a\nb...", "a\nb\nc"));
+        assert!(helper("a\n...b...", "a\nb"));
+        assert!(helper("a\n...b...", "a\nxbz"));
+        assert!(helper("a\n...b...", "a\nbz"));
+        assert!(helper("a\n...b...", "a\nxb"));
+        assert!(!helper("a\n...b...", "a\nxb\nc"));
+    }
+
+    #[test]
+    fn name_matcher() {
+        let nameptn_re = Regex::new("\\$.+?\\b").unwrap();
+        let name_re = Regex::new(".+?\\b").unwrap();
+        let helper = |ptn: &str, text: &str| -> bool {
+            FMBuilder::new(ptn)
+                .unwrap()
+                .name_matcher(Some((nameptn_re.clone(), name_re.clone())))
+                .build()
+                .unwrap()
+                .matches(text)
+                .is_ok()
+        };
+        assert!(helper("", ""));
+        assert!(helper("a", "a"));
+        assert!(!helper("a", "ab"));
+        assert!(helper("...", ""));
+        assert!(helper("...", "a"));
+        assert!(helper("...", "a\nb"));
+        assert!(helper("...\na", "a"));
+        assert!(helper("...\na\n...", "a"));
+        assert!(helper("a\n...", "a"));
+        assert!(helper("a\n...\nd", "a\nd"));
+        assert!(helper("a\n...\nd", "a\nb\nc\nd"));
+        assert!(!helper("a\n...\nd", "a\nb\nc"));
+        assert!(helper("a\n...\nc\n...\ne", "a\nb\nc\nd\ne"));
+        assert!(helper("a\n...\n...b", "a\nb"));
+        assert!(helper("a\n...\nb...", "a\nb"));
+        assert!(helper("a\n...\nb...", "a\nbc"));
+        assert!(helper("a\nb...", "a\nbc"));
+        assert!(!helper("a\nb...", "a\nb\nc"));
+        assert!(helper("a\n...b...", "a\nb"));
+        assert!(helper("a\n...b...", "a\nxbz"));
+        assert!(helper("a\n...b...", "a\nbz"));
+        assert!(helper("a\n...b...", "a\nxb"));
+        assert!(!helper("a\n...b...", "a\nxb\nc"));
```

Comment:
Let's add a test for `......`? As I understand, it's the same as `...` on its own line?

Reply:
Agreed that tests are useful. I think that `......` should be different than `...` in that it should mean "match exactly one line, no more, no less". That just so happens to be the existing code's behaviour ;) See https://github.com/softdevteam/fm/pull/1/commits/c668ee7233b235e1122661fe1064376a8cba199d.

Reply:
You're right! And that may even be a useful thing too.